### PR TITLE
Remove Tpetra::Distributor dependency of Search package

### PR DESCRIPTION
**Pull request by dalg24**
Also replace Teuchos::Comm by raw MPI communicator in the interface of the distributed search tree

The interesting part is in the definition of `Details::Distributor` that is replacing `Tpetra::Distributor`.  For now I decided to keep the same interface.

The PR is marked as "work in progress" because I still wanted to revisit the `extract_and_sort_ranks()` routine but reviews and comments are welcome at this time.

### CODE REVIEW

File: packages/Search/src/details/DTK_DetailsDistributor.hpp [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+// * replace parallel_reduce with std::max_element because I couldn't get it to compile
+// * append size to offsets in main driver but might want to handle it here
+static void extract_and_sort_ranks(
+    Kokkos::View<int*, Kokkos::HostSpace> destination_ranks,
+    Kokkos::View<int*, Kokkos::HostSpace> permutation,
+    std::vector<int>& unique_ranks,
+    std::vector<int>& offsets,
+    std::vector<int>& counts) {
+  int mpi_rank;
+  MPI_Comm_rank(MPI_COMM_WORLD, &mpi_rank);
```
Comment (dalg24) [inline packages/Search/src/details/DTK_DetailsDistributor.hpp]: Do not use `MPI_COMM_WORLD`

File: packages/Search/src/details/DTK_DetailsDistributor.hpp [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+ ****************************************************************************/
+#ifndef DTK_DETAILS_DISTRIBUTOR_HPP
+#define DTK_DETAILS_DISTRIBUTOR_HPP
+
+#include <Kokkos_Core.hpp> // FIXME
+
+#include <mpi.h>
+
+#include <algorithm>
+#include <cassert>
```
Comment (dalg24) [inline packages/Search/src/details/DTK_DetailsDistributor.hpp]: `#include <numeric>`

File: packages/Search/src/details/DTK_DetailsDistributedSearchTreeImpl.hpp [hunk 3]
```diff
@@ -199,12 +192,18 @@ DistributedSearchTreeImpl<DeviceType>::sendAcrossNetwork(
...
-            exports_host.data(), exports_host.size() ),
-        num_packets,
-        Teuchos::ArrayView<typename View::non_const_value_type>(
-            imports_host.data(), imports_host.size() ) );
+    using NonConstValueType = typename View::non_const_value_type;
+    using ConstValueType = typename View::const_value_type;
+
+    Kokkos::View<ConstValueType *, Kokkos::HostSpace,
+                 Kokkos::MemoryTraits<Kokkos::Unmanaged>>
+        export_buffer( exports_host.data(), exports_host.size() );
```
Comment (Rombur) [inline packages/Search/src/details/DTK_DetailsDistributedSearchTreeImpl.hpp]: Is this a shallow or a deep copy?

File: packages/Search/src/details/DTK_DetailsDistributor.hpp [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+    //Kokkos::parallel_reduce("find next biggest rank", n, KOKKOS_LAMBDA(ST i, int& local_max) {
+    //  auto r = tmp_ranks(i);
+    //  local_max = (r > local_max) ? r : local_max;
+    //}, Kokkos::Experimental::Max<int, Kokkos::HostSpace>(next_biggest_rank));
+    next_biggest_rank = *std::max_element(tmp_ranks.data(), tmp_ranks.data() + n);
+    if (next_biggest_rank == -1) break;
+    unique_ranks.push_back(next_biggest_rank);
+    offsets.push_back(offset);
+    Kokkos::View<int, Kokkos::HostSpace> total("total");
+    Kokkos::parallel_scan("process biggest rank items", Kokkos::RangePolicy<Kokkos::Serial>(0, n),
```
Comment (Rombur) [inline packages/Search/src/details/DTK_DetailsDistributor.hpp]: I am confused are you doing the sort in serial? Then why not use STL

File: packages/Search/src/details/DTK_DetailsDistributor.hpp [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+        if (i + 1 == tmp_ranks.extent(0)) {
+          total() = index;
+        }
+        if (tmp_ranks(i) == next_biggest_rank) {
+          tmp_ranks(i) = -1;
+        }
+      }
+    });
+    auto host_total = Kokkos::create_mirror_view_and_copy(
+        Kokkos::HostSpace(), total);
```
Comment (Rombur) [inline packages/Search/src/details/DTK_DetailsDistributor.hpp]: Isn't `total` already on the host?

File: packages/Search/src/details/DTK_DetailsDistributor.hpp [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+    Kokkos::View<int, Kokkos::HostSpace> total("total");
+    Kokkos::parallel_scan("process biggest rank items", Kokkos::RangePolicy<Kokkos::Serial>(0, n),
+    KOKKOS_LAMBDA(ST i, int& index, const bool last_pass) {
+      if (last_pass && (tmp_ranks(i) == next_biggest_rank)) {
+        permutation(i) = index + offset;
+      }
+      if (tmp_ranks(i) == next_biggest_rank) ++index;
+      if (last_pass) {
+        if (i + 1 == tmp_ranks.extent(0)) {
+          total() = index;
```
Comment (Rombur) [inline packages/Search/src/details/DTK_DetailsDistributor.hpp]: What does `total()` do?

File: packages/Search/src/details/DTK_DetailsDistributor.hpp [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+class Distributor
+{
+  public:
+    Distributor( MPI_Comm comm )
+        : _comm( comm )
+    {
+    }
+
+    template <typename T>
+    static T *notNullPtr( T *p )
```
Comment (Rombur) [inline packages/Search/src/details/DTK_DetailsDistributor.hpp]: What's the point of that function?

File: packages/Search/src/details/DTK_DetailsDistributor.hpp [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+                _comm, 1, sources, degrees, notNullPtr( destinations.data() ),
+                MPI_UNWEIGHTED, MPI_INFO_NULL, reorder, &_comm_dist_graph );
+        }
+
+        int indegrees;
+        int outdegrees;
+        int weighted;
+        MPI_Dist_graph_neighbors_count( _comm_dist_graph, &indegrees,
+                                        &outdegrees, &weighted );
+        assert( weighted == 0 );
```
Comment (Rombur) [inline packages/Search/src/details/DTK_DetailsDistributor.hpp]: Use `DTK_REQUIRE` instead of `assert`

File: packages/Search/src/details/DTK_DetailsDistributor.hpp [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+        assert( outdegrees == (int)destinations.size() );
```
Comment (Rombur) [inline packages/Search/src/details/DTK_DetailsDistributor.hpp]: Use `static_cast`

File: packages/Search/src/details/DTK_DetailsDistributor.hpp [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+            if ( it != sources.end() )
+            {
+                int const j = std::distance( sources.begin(), it );
+                std::iota( &_permute_recv[offset],
+                           &_permute_recv[offset] + _src_counts[j],
+                           _src_offsets[j] );
+                offset += _src_counts[j];
+            }
+        }
+        assert( offset == (int)_permute_recv.size() );
```
Comment (Rombur) [inline packages/Search/src/details/DTK_DetailsDistributor.hpp]: `assert` and `static_cast`

File: packages/Search/src/details/DTK_DetailsDistributor.hpp [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+#endif
+
+        return _src_offsets.back();
+    }
+    template <typename View>
+    void doPostsAndWaits( typename View::const_type const &exports,
```
Comment (Rombur) [inline packages/Search/src/details/DTK_DetailsDistributor.hpp]: Is this the name we want to stick with? I find `doPostsAndWaits` a strange name for a function.

File: packages/Search/src/details/DTK_DetailsDistributor.hpp [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+                           &_permute_recv[offset] + _src_counts[j],
+                           _src_offsets[j] );
+                offset += _src_counts[j];
+            }
+        }
+        assert( offset == (int)_permute_recv.size() );
+#endif
+
+        return _src_offsets.back();
+    }
```
Comment (Rombur) [inline packages/Search/src/details/DTK_DetailsDistributor.hpp]: Add a blank line

File: packages/Search/src/details/DTK_DetailsDistributor.hpp [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+
+        assert( (size_t)src_offsets.back() ==
+                imports.size() * sizeof( ValueType ) );
+        assert( (size_t)dest_offsets.back() ==
+                exports.size() * sizeof( ValueType ) );
+
+        for ( int i = 0; i < _dest_offsets.back(); ++i )
+            std::copy( &exports[numPackets * i],
+                       &exports[numPackets * i] + numPackets,
+                       &dest_buffer[numPackets * _permute[i]] );
```
Comment (Rombur) [inline packages/Search/src/details/DTK_DetailsDistributor.hpp]: How does this work when `exports` in on the device?

File: packages/Discretization/src/DTK_PointSearch_def.hpp [hunk 14]
```diff
@@ -88,7 +89,9 @@ PointSearch<DeviceType>::PointSearch(
     Teuchos::RCP<const Teuchos::Comm<int>> comm, Mesh<DeviceType> const &mesh,
     Kokkos::View<double **, DeviceType> points_coordinates )
     : _comm( comm )
-    , _target_to_source_distributor( _comm )
+    , _target_to_source_distributor(
+          *( Teuchos::rcp_dynamic_cast<Teuchos::MpiComm<int> const>( comm )
```
[the remark this answers is not in the document]
Comment (dalg24) [reply packages/Discretization/src/DTK_PointSearch_def.hpp]: OK. I actually wanted to check with you whether you want to change the constructor of the interpolation operator to propagate the changes even further.

File: packages/Search/src/details/DTK_DetailsDistributedSearchTreeImpl.hpp (hunk 3 above)
Comment (dalg24) [reply packages/Search/src/details/DTK_DetailsDistributedSearchTreeImpl.hpp]: This is not a copy, this "casts" the argument into a rank-1 view.

File: packages/Search/src/details/DTK_DetailsDistributor.hpp (hunk 4 above)
Comment (dalg24) [reply packages/Search/src/details/DTK_DetailsDistributor.hpp]: Copy/paste from Kokkos tutorial.  Will revisit.

File: packages/Search/src/details/DTK_DetailsDistributor.hpp (hunk 7 above)
Comment (dalg24) [reply packages/Search/src/details/DTK_DetailsDistributor.hpp]: I need to think about the name a bit more carefully and comment but in short, OpenMPI routines assert that some of the pointers passed as argument are not null even when no data will be communicated (`std::vector::data()` returns `nullptr` when the container is empty)

File: packages/Search/src/details/DTK_DetailsDistributor.hpp (hunk 8 above)
Comment (dalg24) [reply packages/Search/src/details/DTK_DetailsDistributor.hpp]: Will do

File: packages/Search/src/details/DTK_DetailsDistributor.hpp (hunk 11 above)
Comment (dalg24) [reply packages/Search/src/details/DTK_DetailsDistributor.hpp]: I decided to stick with the Tpetra interface for now but intend on revisiting later.  For instance, I don't really like that the class stores two MPI communicators.  I might turn `createFromSends()` into a static member function that builds the `Distributor` which essentially becomes a wrapper for send and receive counts/offsets with a method for executing the communication plan.

File: packages/Search/src/details/DTK_DetailsDistributor.hpp (hunk 9 above)
Comment (dalg24) [reply packages/Search/src/details/DTK_DetailsDistributor.hpp]: Can do.  Just out of curiosity, why do you prefer the `static_cast`?

File: packages/Search/src/details/DTK_DetailsDistributor.hpp (hunk 13 above)
Comment (dalg24) [reply packages/Search/src/details/DTK_DetailsDistributor.hpp]: (looks like my reply has been lost...)
I want to add a static assertion that the memory space is accessible from the host.  They are still a few rough edges with `sendAcrossNetwork()` and the distributor construction that need to be reworked (e.g. unnecessary copies)

File: packages/Search/src/details/DTK_DetailsDistributor.hpp (hunk 6 above)
Comment (dalg24) [reply packages/Search/src/details/DTK_DetailsDistributor.hpp]: Data access for a rank-0 view

File: packages/Search/src/details/DTK_DetailsDistributor.hpp (hunk 7 above)
Comment (dalg24) [reply packages/Search/src/details/DTK_DetailsDistributor.hpp]: I added a note above the function declaration to capture this.

File: packages/Search/src/details/DTK_DetailsDistributor.hpp (hunk 13 above)
Comment (dalg24) [reply packages/Search/src/details/DTK_DetailsDistributor.hpp]: I added a static assertion that the underlying data in the view is accessible from the host.

**Comment by dalg24** [comment]
Ready for review.  I am not sure what's going on with the timeouts.

File: packages/Discretization/test/tstInterpolation.cpp [hunk 15]
```diff
@@ -273,7 +273,9 @@ TEUCHOS_UNIT_TEST_TEMPLATE_1_DECL( Interpolation, one_topo_one_fe_three_dim,
     DataTransferKit::Mesh<DeviceType> mesh( cell_topologies, cells,
                                             coordinates );
     DataTransferKit::Interpolation<DeviceType> interpolation(
-        comm, mesh, points_coord, cell_dofs_ids, DTK_HGRAD );
+        *( Teuchos::rcp_dynamic_cast<Teuchos::MpiComm<int> const>( comm )
+               ->getRawMpiComm() ),
```
Comment (Rombur) [inline packages/Discretization/test/tstInterpolation.cpp]: There is no reason to build a Teuchos communicator. Just create a raw mpi communicator

File: packages/Discretization/test/tstPointSearch.cpp [hunk 16]
```diff
@@ -173,8 +173,10 @@ TEUCHOS_UNIT_TEST_TEMPLATE_1_DECL( PointSearch, one_topo_three_dim, DeviceType )
 
     DataTransferKit::Mesh<DeviceType> mesh( cell_topologies_view, cells,
                                             coordinates );
-    DataTransferKit::PointSearch<DeviceType> pt_search( comm, mesh,
-                                                        points_coord );
+    DataTransferKit::PointSearch<DeviceType> pt_search(
+        *( Teuchos::rcp_dynamic_cast<Teuchos::MpiComm<int> const>( comm )
+               ->getRawMpiComm() ),
```
Comment (Rombur) [inline packages/Discretization/test/tstPointSearch.cpp]: Just use a raw mpi communicator from the start

File: packages/Search/src/details/DTK_DetailsDistributor.hpp [hunk 17]
```diff
@@ -28,65 +28,63 @@ namespace DataTransferKit
...
+        return;
+
+    Kokkos::View<int *, Kokkos::HostSpace> ranks_duplicate(
+        Kokkos::ViewAllocateWithoutInitializing( ranks.label() ),
+        ranks.size() );
+    Kokkos::deep_copy( ranks_duplicate, ranks );
+
+    while ( true )
+    {
+        // TODO consider replacing with parallel reduce
```
Comment (Rombur) [inline packages/Search/src/details/DTK_DetailsDistributor.hpp]: Do you have any plan to change this or should we just remove this code and use `std::sort`

File: packages/Search/src/details/DTK_DetailsDistributor.hpp [hunk 18]
```diff
@@ -112,7 +112,7 @@ class Distributor
     {
         static_assert(
             std::is_same<typename View::non_const_value_type, int>::value, "" );
-        int comm_rank = -1;
+        int comm_rank;
```
Comment (Rombur) [inline packages/Search/src/details/DTK_DetailsDistributor.hpp]: Why did you change this?

File: docker/jenkins/build.sh [hunk 19]
```diff
@@ -39,7 +39,7 @@ fi
 # build
 make -j${NPROC}
 # run the unit tests
-ctest -j${NPROC} --no-compress-output -T Test
+ctest -j1 --no-compress-output -T Test
```
Comment (dalg24) [inline docker/jenkins/build.sh]: @Rombur This is all I could come up with for now

I think the timeout are related to #375

Comment (Rombur) [reply docker/jenkins/build.sh]: I don't see how that makes sense that #375 is the problem. If you need to do that it means we cannot run two instances of DTK at the same time. So it could be that there is something that's acting like global variable in the code.

Comment (dalg24) [reply docker/jenkins/build.sh]: Trying to set the `OMP_NUM_THREADS` env variable instead...

File: packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator_def.hpp [hunk 20]
```diff
@@ -13,13 +13,14 @@
...
 #include <DTK_DetailsDistributedSearchTreeImpl.hpp>
+#include <DTK_DetailsDistributor.hpp>
 #include <DTK_DetailsUtils.hpp>
 
 #include <netcdf.h>
 
 #include <Teuchos_Array.hpp>
 
-#include <Tpetra_Distributor.hpp>
+#include <Tpetra_Distributor.hpp> // FIXME???
```
Comment (sslattery) [inline packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator_def.hpp]: Why is this include still needed?

File: packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator_def.hpp [hunk 21]
```diff
@@ -351,7 +355,7 @@ void ExodusProblemGenerator<Scalar, SourceDevice, TargetDevice>::
     }
 
     // Build a communication plan for the sources.
-    Tpetra::Distributor distributor( _comm );
+    Tpetra::Distributor distributor( _comm ); // FIXME???
```
Comment (sslattery) [inline packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator_def.hpp]: Can we use your distributor here?

File: packages/Search/src/DTK_DistributedSearchTree_decl.hpp [hunk 22]
```diff
@@ -107,27 +103,25 @@ class DistributedSearchTree
...
+    MPI_Comm_rank( _comm, &comm_rank );
+    int comm_size;
+    MPI_Comm_size( _comm, &comm_size );
 
     Kokkos::View<Box *, DeviceType> boxes(
         Kokkos::ViewAllocateWithoutInitializing( "rank_bounding_boxes" ),
         comm_size );
-    // FIXME: I am not sure how to do the MPI allgather with Teuchos for data
-    // living on the device so I copied to the host.
+    // FIXME when we move to MPI with CUDA-aware support, we will not need to
```
Comment (sslattery) [inline packages/Search/src/DTK_DistributedSearchTree_decl.hpp]: Is this just a matter of setting this up on the CI? Only open-mpi seems to support this right now.

File: packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator_def.hpp (hunk 21 above)
Comment (dalg24) [reply packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator_def.hpp]: For no good reason.  I thought you were using the `Teuchos::Comm` but it seems you only get the size and the rank...

File: packages/Search/src/DTK_DistributedSearchTree_decl.hpp (hunk 22 above)
Comment (dalg24) [reply packages/Search/src/DTK_DistributedSearchTree_decl.hpp]: We would probably need to check at configure time that CUDA support is enabled.

File: packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator_def.hpp (hunk 20 above)
Comment (dalg24) [reply packages/Meshfree/test/PointCloudProblemGenerator/ExodusProblemGenerator_def.hpp]: I remember now, because I did not want us to call `Details::Distributor::doPostsAndWaits()` directly

File: packages/Search/src/DTK_DistributedSearchTree_decl.hpp (hunk 22 above)
Comment (Rombur) [reply packages/Search/src/DTK_DistributedSearchTree_decl.hpp]: All the main mpi implementations support cuda.

File: packages/Search/src/details/DTK_DetailsDistributor.hpp [hunk 23]
```diff
@@ -0,0 +1,252 @@
...
+                               _src_counts.data(), 1, MPI_INT,
+                               _comm_dist_graph );
+
+        _src_offsets.resize( indegrees + 1 );
+        // exclusive scan
+        _src_offsets[0] = 0;
+        for ( int i = 0; i < indegrees; ++i )
+            _src_offsets[i + 1] = _src_offsets[i] + _src_counts[i];
+
+#if defined( REORDER_RECV )
```
Comment (sslattery) [inline packages/Search/src/details/DTK_DetailsDistributor.hpp]: So is this to reorder the output of the graph communication into a preferred order?

File: packages/Search/src/details/DTK_DetailsDistributor.hpp [hunk 24]
```diff
@@ -0,0 +1,252 @@
...
+
+        // TODO
+        // * apply permutation on the device in a parallel for
+        // * switch to MPI with CUDA support (do not copy to host)
+        for ( int i = 0; i < _dest_offsets.back(); ++i )
+            std::copy( &exports[num_packets * i],
+                       &exports[num_packets * i] + num_packets,
+                       &dest_buffer[num_packets * _permute[i]] );
+
+        MPI_Neighbor_alltoallv(
```
Comment (sslattery) [inline packages/Search/src/details/DTK_DetailsDistributor.hpp]: Yeah this is so much easier now that I see all of this. We will definitely want to do some performance tests in Cabana to see which way performs better. This way is certainly cleaner and more maintainable.

File: packages/Search/src/details/DTK_DetailsDistributor.hpp [hunk 25]
```diff
@@ -0,0 +1,252 @@
...
+    void doPostsAndWaits( typename View::const_type const &exports,
+                          size_t num_packets, View const &imports ) const
+    {
+        DTK_REQUIRE( num_packets * _src_offsets.back() == imports.size() );
+        DTK_REQUIRE( num_packets * _dest_offsets.back() == exports.size() );
+
+        using ValueType = typename View::value_type;
+        static_assert( View::rank == 1, "" );
+        static_assert(
+            Kokkos::Impl::MemorySpaceAccess<typename View::memory_space,
```
Comment (sslattery) [inline packages/Search/src/details/DTK_DetailsDistributor.hpp]: So does this assume that when this is used elsewhere in the code that we will always have moved data to the host before this class is called?

File: packages/Search/src/details/DTK_DetailsDistributor.hpp [hunk 26]
```diff
@@ -0,0 +1,252 @@
...
+        MPI_Comm_rank( _comm, &comm_rank );
+
+        _permute = Kokkos::View<int *, Kokkos::HostSpace>(
+            Kokkos::ViewAllocateWithoutInitializing( "permute" ),
+            destination_ranks.size() );
+        std::vector<int> destinations;
+        sortAndDetermineBufferLayout( destination_ranks, _permute, destinations,
+                                      _dest_counts, _dest_offsets );
+
+        {
```
[the remark this answers is not in the document]
Comment (dalg24) [reply packages/Search/src/details/DTK_DetailsDistributor.hpp]: Because the two C arrays are not needed after we made the new communicator and I want to reclaim the name `sources`.

File: packages/Search/src/details/DTK_DetailsDistributor.hpp [hunk 27]
```diff
@@ -0,0 +1,252 @@
...
+            int const reorder = 0; // Ranking may *not* be reordered
+            int const sources[1] = {comm_rank};
```
[the remark this answers is not in the document]
Comment (dalg24) [reply packages/Search/src/details/DTK_DetailsDistributor.hpp]: > So you only have one source in this call to `MPI_Dist_graph_create` because we only know where this rank is sending right?

Correct

> I wonder if it will resolve collisions where the same edge is input on multiple ranks.

It is ok to list the same edge on different processes.  It will only appears once in the resulting communication graph.  I am not sure how it would handle weights in that case though.

File: packages/Search/src/details/DTK_DetailsDistributor.hpp (hunk 23 above)
Comment (dalg24) [reply packages/Search/src/details/DTK_DetailsDistributor.hpp]: Reordering the receive buffer.  Some of the downstream code (by that I mean packages other than `Search`) seems to be sensitive to it.  This needs more investigation.  My goal is to get rid of it.

File: packages/Search/src/details/DTK_DetailsDistributor.hpp (hunk 25 above)
Comment (dalg24) [reply packages/Search/src/details/DTK_DetailsDistributor.hpp]: It does not assume anything.  It will not compile if the data is not accessible from the host.  I do not intend anyone to call `Details::Distributor::doPostsAndWaits()` directly, instead one should be using `sendAcrossNetwork(distributor, export, import)`.  I intend to change the interface of the `Distributor` and "merge" these two functions into a new member function.  It is beside the scope of this PR though.

File: packages/Search/src/details/DTK_DetailsDistributor.hpp [hunk 28]
```diff
@@ -0,0 +1,252 @@
...
+        MPI_Dist_graph_neighbors_count( _comm_dist_graph, &indegrees,
+                                        &outdegrees, &weighted );
+        DTK_ENSURE( weighted == 0 );
+        DTK_ENSURE( outdegrees == static_cast<int>( destinations.size() ) );
+
+        std::vector<int> sources( indegrees );
+        MPI_Dist_graph_neighbors( _comm_dist_graph, indegrees,
+                                  notNullPtr( sources.data() ), MPI_UNWEIGHTED,
+                                  outdegrees, notNullPtr( destinations.data() ),
+                                  MPI_UNWEIGHTED );
```
Comment (aprokop) [inline packages/Search/src/details/DTK_DetailsDistributor.hpp]: Is there a guarantee that the neighbors are going to be listed in the same order as original `destinations`? Could not find this anywhere in the documentation. In other words, could this call permute `destinations`? If yes, then the next `MPI_Neighbor_alltoall` may do the wrong thing.

Comment (dalg24) [reply packages/Search/src/details/DTK_DetailsDistributor.hpp]: > Is there a guarantee that the neighbors are going to be listed in the same order as original `destinations`? 

No

> If yes, then the next `MPI_Neighbor_alltoall` may do the wrong thing.

I don't see how "permute destinations" would make  `MPI_Neighbor_alltoall` do the wrong thing

Comment (dalg24) [reply packages/Search/src/details/DTK_DetailsDistributor.hpp]: `destination` is an output parameter (I realize this is confusing because it overwrite whatever was in the vector until this point)
This is precisely how MPI tells you how receives are going to be laid off.

Comment (aprokop) [reply packages/Search/src/details/DTK_DetailsDistributor.hpp]: Talked to @dalg24 about the thing, and he is going to investigate this.

File: packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp [hunk 29]
```diff
@@ -204,7 +204,9 @@ void testUniquelyOwnedProblem(
 
     // Create a nearest neighbor operator.
     DataTransferKit::NearestNeighborOperator<Device> nearest_op(
-        comm, src_coords_device, tgt_coords_device );
+        *( Teuchos::rcp_dynamic_cast<Teuchos::MpiComm<int> const>( comm )
+               ->getRawMpiComm() ),
```
Comment (Rombur) [inline packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp]: Do we use `Teuchos::MpiComm` in this test?

File: packages/Search/examples/distributed_tree_driver/distributed_tree_driver.cpp [hunk 30]
```diff
@@ -164,7 +165,9 @@ int main_( Teuchos::CommandLineProcessor &clp, int argc, char *argv[] )
     comm->barrier();
     construction->start();
     DataTransferKit::DistributedSearchTree<DeviceType> distributed_tree(
-        comm, bounding_boxes );
+        *( Teuchos::rcp_dynamic_cast<Teuchos::MpiComm<int> const>( comm )
+               ->getRawMpiComm() ),
```
Comment (Rombur) [inline packages/Search/examples/distributed_tree_driver/distributed_tree_driver.cpp]: Why not using a raw MPI communicator?

File: packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp (hunk 29 above)
Comment (dalg24) [reply packages/Meshfree/test/tstNearestNeighborExodusGenerator.cpp]: The test use free functions from `Teuchos_CommHelpers.hpp` and builds a `Tpetra::Map`
